Add Scott's Allies02 multiplayer mission as "Evacuation"

obrakmann opened the pull request:
Converted from the [C# mission script found in release-20131223](https://github.com/OpenRA/OpenRA/blob/release-20131223/OpenRA.Mods.RA/Missions/Allies02Script.cs).

It follows the original script pretty closely, there are only a few minor deviations. That also means that the difficulty is quite high, and I'll probably end up introducing a new "Easy" skill level and shifting the other two up accordingly. I'll need some feedback on that, however.

// cc: @ScottNZ

obrakmann committed:
Fix winstate logic for cooperative games

obrakmann committed:
Add Scott's Allies02 multiplayer mission as "Evacuation"

obrakmann committed:
Add another way to call air strikes/paradrops to the Lua API

obrakmann committed:
Add a Lua API method to randomize a collection

obrakmann committed:
Add methods to retrieve player stats to Lua API

obrakmann committed:
Add missing `Requires` to MissionObjectiveProperties

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+	soviets.Cash = 1000
+
+	if Map.Difficulty == "Easy" then
+		for i, sam in ipairs(Sams) do
```
Why didn't you use `Utils.Do` here?

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+end
+
+ReassignActors = function(actors, from, to)
+	for i, a in ipairs(actors) do
```
Same here: `Utils.Do(actors, function(a)...`

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+
+	soviets = Player.GetPlayer("Soviets")
+
+	for i, player in ipairs({ allies1, allies2 }) do
```
And same question here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs:
```diff
@@ -215,7 +215,10 @@ public void OnPlayerLost(Player player)
 					if (Info.EarlyGameOver)
 						foreach (var p in enemies)
-							p.PlayerActor.Trait<MissionObjectives>().ForceDefeat(p);
+						{
```
Doesn't our code style also require braces around the `if` now?

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+Paratroopers =
+{
+	{
+		-- `proxy` cannot be assigned in global scope since a player is needed to create the actor
```
Why do we need to have the actor inside this array anyway?
I think using the name would be better:
```
Paratroopers =
{
	{
		proxy = "powerproxy.paras1",
		entry = BadgerEntryPoint1.Location,
		drop  = BadgerDropPoint1.Location,
	},
[...]
```
And then in `SendParatroopers` (as we use this only once anyway):
```
Utils.Do(Paratroopers, function(para)
	local units = Actor.Create(para.proxy, false, { Owner = soviets }).SendParatroopersFrom(para.entry, para.drop)
	Utils.Do(units, function(unit)
[...]

abcdefg30 replied:
Ah wait, we might want to destroy the proxy actor as we don't use it again anyway, so I'd suggest following instead:
```
Utils.Do(Paratroopers, function(para)
    local proxy = Actor.Create(para.proxy, false, { Owner = soviets })
    local units = proxy.SendParatroopersFrom(para.entry, para.drop)
    proxy.Destroy()
    Utils.Do(units, function(unit)
[...]

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+		proxy = nil,
+		entry = BadgerEntryPoint1.Location + CVec.New(6, 0),
+		drop  = BadgerDropPoint3.Location,
+	},
```
I'm not sure what our convention is here. I never add a comma behind the last element of an array. That is up to you though, I won't really complain about it (much). I just wondered as this is even inconsistently done in this single script.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+	yak.Move(Map.ClosestEdgeCell(yak.Location))
+	yak.Destroy()
+	Trigger.OnRemovedFromWorld(Yak, function()
+		Yak = nil
```
Hm, I don't think we really need this, we could just add an `Yak.IsInWorld` check at the `OnEnteredProximityTrigger` instead: `if Yak.IsInWorld or actor.Owner ~= allies1 `

obrakmann replied:
I'll keep it this way since it's easier to code. `Yak` would still be nil when the trigger first fires, so I need to check for it anyway.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+	local proxy = Actor.Create("powerproxy.parabombs", false, { Owner = soviets })
+	proxy.SendAirstrikeFrom(BadgerEntryPoint2.Location, ParabombPoint1.Location)
+	proxy.SendAirstrikeFrom(BadgerEntryPoint2.Location + CVec.New(0, 3), ParabombPoint2.Location)
+end
```
Similiar thing here: Just destroy the proxy actor when we don't need it anymore.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+
+	Utils.Do(AttackGroup, function(unit)
+		if not unit.IsDead then
+			unit.Hunt()
```
I guess you *could* remove this if you feel like, as you assign an "IdleHunt" trigger to the unit the line below anyway.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+	end
+
+	soviets.Build({ Utils.Random(SovietInfantry) }, function(units)
+		unit = units[1]
```
`local unit` please, but I'd drop that entirely and just use `table.insert(AttackGroup, units[1])` instead below.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+	end
+
+	soviets.Build({ Utils.Random(SovietVehicles[SovietVehicleType]) }, function(units)
+		unit = units[1]
```
Same here.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+		end
+	end
+
+	if allies2 == nil or allies2.IsLocalPlayer then
```
We usually don't explicitly check for `nil`. I'd just use `if not allies2 or allies2.IsLocalPlayer then`.

abcdefg30 commented on mods/ra/maps/evacuation/evacuation.lua:
```diff
@@ -0,0 +1,365 @@
+		Camera.Position = ChinookHusk.CenterPosition
+	end
+
+	if allies2 == nil then
```
Same here. (`if not allies2 then`)

abcdefg30 commented on mods/ra/maps/evacuation/rules.yaml:
```diff
@@ -0,0 +1,182 @@
+
+World:
+	MissionData:
+		Briefing: One of our outposts is under heavy attack by a nearby Soviet base. Einstein, Tanya and many of our forces have been evacuated via helicopters, but those were shot down on the other side of the river. Fortunately they survived the crash, but are now in enemy territory. Tanya and the rest of the surviving forces need to fight their way out of there. The air defenses must be eliminated b ...
```
Could you add another `\n` at the end, please? This increases readability:
![evacuationdesca](https://cloud.githubusercontent.com/assets/7704140/15378321/dd91cc66-1d62-11e6-8d6c-378e48bcd541.png)
vs
![evacuationdescb](https://cloud.githubusercontent.com/assets/7704140/15378326/e32aa06c-1d62-11e6-973e-cfc7d71c4a01.png)

abcdefg30 commented:
Some issues I noticed during my first playthrough:
- You can see following units but will never be able to get their prerequisites: Gap Generator, Chronosphere, Iron Curtain, Missile Silo, M.A.D Tank, MiG, Longbow, Missile Sub
- An exception from above is the Flamethrower. You can - in theory - get the required prerequisite by capturing a flame tower... I think here we should just make flame towers buildable.
- Just minor: It feels a bit weird that Heavy Tanks require a service depot and Medium Tanks don't.
- Einstein has the same selection priority as normal soldiers while the engineer has not, this is somehow confusing imho.
- If you play alone you get "Mission Accomplished", but the game just continues.
- Just for bonus points: I had to search a bit before I found the flare and the helicopter I need to send Einstein to. What about placing a beacon there?
- To "That also means that the difficulty is quite high [...]": Well, not too, imho. It was quite easy to snipe ever soldier with tanya as they don't come for you like they do in the allies03 missions, you could add that for extra bonus points. :D

obrakmann committed:
Fix crash in Beacon ctor when it runs during instead of between ticks

Causes an exception like below, for example when calling the Lua API's Beacon.New() method.

``Exception of type `System.InvalidOperationException`: Collection was modified; enumeration operation may not execute.``

obrakmann commented:
Updated.

Re: the points in your last comments:

1. Fixed.
2. Made FTUR buildable.
3. Yeah. Do you think I should add a FIX to the base?
4. I would very much like to revert the selection priority change for the engineer, it's totally annoying. Shouldn't be part of this PR, though.
5. That was due to the dummy 'Allies' actor not being marked as a non-combatant, I'm pretty sure.
6. Beacon added.
7. Yeah, no. You're right, on easy it's quite doable.

abcdefg30 commented:
> 3. Yeah. Do you think I should add a FIX to the base?

Hm, I'm not really convinced. Would that work good together with power management? And it would make it easier to not loose 100 units. So it' rather say don't add a FIX.

> 7. Yeah, no. You're right, on easy it's quite doable.

I talked about normal, but ok. :p

obrakmann commented:
Forgot to lock the tech level and debug lobby options previously, fixed now.

cjshmyr commented:
Haven't checked the code, but map played well :+1:

Mailaender commented:
Nicely done. Works well (singleplayer) :white_check_mark:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/62f1cfcba470dca435f3928bf6a73f6c6d32d086...d752175cacec624e46ca4dbb35c118b1713c21c7)

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/62f1cfcba470dca435f3928bf6a73f6c6d32d086...d752175cacec624e46ca4dbb35c118b1713c21c7)

pchote commented on mods/ra/maps/evacuation/map.yaml:
```diff
@@ -0,0 +1,1757 @@
+
+Bounds: 16,16,96,96
+
+Visibility: MissionSelector, Lobby
```
Maps should be one or the other, not both.

Mailaender replied:
These are Co-Op mission which can be played single-player as well so this entry is valid. Our UI is just not capable enough. There is even a TODO in the code where it crashes.

pchote replied:
Having a thing appear in two different places that otherwise display completely disjoint content is terrible UX.

Mailaender replied:
For me the way to solve https://github.com/OpenRA/OpenRA/issues/11692 is to
* add a multi-selection dropdown
* disable UI support for the setting
* make it a simple text edit widget

I wouldn't recommend just avoid using values and I don't see how having this on both the skirmish lobby and the mission selector contributes to negative UX. It just makes it easier to discover.

obrakmann replied:
But it doesn't make sense to duplicate the whole map just with different Visibility entries, either. And the map *is* both single- and multi-player capable. I was very careful to allow for that especially.

Either way, the fix is easy: http://stackoverflow.com/a/8317808

Mailaender commented on mods/ra/maps/evacuation/rules.yaml:
```diff
@@ -0,0 +1,217 @@
+CAMERA.Large:
+	Inherits: CAMERA
+	RevealsShroud:
+		Range: 1000c0
```
This would trigger:

```
Exception of type `System.ArgumentOutOfRangeException`: The requested range (101) cannot exceed the value of MaximumTileSearchRange (50)
Parameter name: maxRange
```

if actually used in-game.

obrakmann replied:
It's not used, it's just [a left-over from the original map](https://github.com/OpenRA/OpenRA/blob/release-20131223/mods/ra/maps/allies-02/map.yaml#L1932-L1935). Also, this PR is merged, no need to review ;)

Mailaender replied:
I just left a note for me and others as I am using your rule port and scripts as templates. https://github.com/OpenRA/OpenRA/pull/11679

Phrohdoh replied:
If it isn't used it should have been removed.

Mailaender replied:
Sorry, the time travel DeLorean is out of fuel.
